README.md
# Perl Quota module

This repository contains the sources of the Perl "Quota" module, which has its
official home at [https://metacpan.org/pod/Quota](https://metacpan.org/pod/Quota).
Please use the ticket interfaces at CPAN for reporting issues.

The Perl Quota module provides access to file system quotas on UNIX platforms.
This works both for locally mounted file systems and network file systems (via
RPC, i.e. Remote Procedure Call). The interface is designed to be independent
of UNIX flavours as well as file system types.

I started developing this module while working as a UNIX system administrator
in 1995 (while being a CS student at university). Since then it was continually
improved and ported to more and more UNIX platforms and filesystems. Numerous
people have contributed to this process; for a complete list of names please
see the CHANGES document in the repository.

The quota module was in the public domain 1995 to 2001. Since 2001 it is
licensed under the same terms as Perl itself, which is (at your choice) either
the Perl Artistic License, or version 1 or later of the GNU General Public
License.  For a copy of these licenses see
<http://www.opensource.org/licenses/>.  The respective authors of the source
code are its owner in regard to copyright.

## Module information

Perl DLSIP-Code: Rcdfg

* stable release
* C compiler required for installation
* support by developer
* plain functions, no references used
* licensed under the Perl Artistic License or (at your option)
  version 1 or later of the GNU General Public License

List of supported operating systems and file systems:

* SunOS 4.1.3
* Solaris 2.4 - 2.10
* HP-UX 9.0x & 10.10 & 10.20 & 11.00
* IRIX 5.2 & 5.3 & 6.2 - 6.5
* OSF/1 & Digital Unix 4
* BSDi 2, FreeBSD 3.x - 4.9, OpenBSD & NetBSD (no RPC)
* Linux - kernel 2.0.30 and later, incl. Quota API v2 and XFS
* AIX 4.1, 4.2 and 5.3
* AFS (Andrew File System) on many of the above (see INSTALL)
* VxFS (Veritas File System) on Solaris 2.

## Documentation

For further information please refer to the following files:

* <A HREF="Quota/Quota.pm">Quota.pm</A>: API documentation (at the end of the file)
* <A HREF="Quota/INSTALL">INSTALL</A>: Installation description
* <A HREF="Quota/CHANGES">CHANGES</A>: Change log &amp; acknowledgements
* <A HREF="Quota/LICENSE">LICENSE</A>: Perl License


Quota/CHANGES
Changes in 1.7.1 (September 2013)
- "make test" now aborts if STDOUT is not a terminal, same as already
  for STDIN. Suggested by Alexandr Ciornii via CPAN ticket 82564.

Changes in 1.7.0 (August 2013)
- Added support for the new NetBSD 6.0 quota system (libquota)
  Thanks to David Holland and Edgar Fuss
- To support limits and usage >= 2^32 on 32-bit clients, Quota::setqlim
  and Quota::query interfaces were changed to use double (NV) instead
  of integer (IV). This should be transparent to users of the module.

Changes in 1.6.7 (November 2011)
- Added support to detect 3.x Linux kernels
  Thanks to Salvatore Bonaccorso of the Debian Perl Group

Changes in 1.6.6 (June 2011)
- Ignore "rootfs" filesystem in Quota::getqcarg() on Linux, which always
  is a duplicate.  Thanks to "abaturin" for providing this patch.
- Made test.pl exit immediately when AUTOMATED_TESTING is set in the
  environment (same as when called not from a tty), as all tests require
  user interaction. Changed on request of TODDR of cpan.org

Changes in 1.6.5 (January 2011)
- Fixed syntax error and missing init for params to setqlim()
  Thanks to Heinrich Mislik for reporting the bugs

Changes in 1.6.4 (January 2010)
- Fixed RPC quota block count truncation to 32bit (CPAN ticket 52813 )
  Thanks to Lloyd Brown for reporting the bug

Changes in 1.6.3 (December 2008)
- Fixed SEG fault for 32-bit Linux clients on top of 64-bit kernels
  Thanks to Jani Ollikainen (pronetko.fi) for reporting and debugging this.
  Thanks to Thomas Rzipa for reporting this issue, too.
- Fixed setqlim to allow setting limits exceeding 2^32 blocks or files.
  The current solution will work only for perl executables which support
  64-bit wide integers (e.g. x86-64)

Changes in 1.6.2 (January 2008)
- Fixed memory leak in RPC authentication
  Thanks again to Kostik (koc at rol.ru)

Changes in 1.6.1 (November 2007)
- Adapted hints/bsd.h to use statvfs instead of statfs in NetBSD 3.0
- Added support for extended quota RPC to allow querying group quotas.
  Thanks to Kostik (koc AT rol.ru) for the suggestion.

Changes in 1.6.0 (October 2007)
- Work-around for different getmntent() semantics for linux loop mounts
  Thanks Wouter de Jong (wouter@widexs.nl) for reporting this issue.
- (late) version bump due to the interface change in 1.5.2

Changes in 1.5.2 (October 2007)
- Added support for new quotactl() parameters in JFS2 on AIX 5.3.
  Thanks to Joshua Frigerio (joshua@uic.edu) for providing a development
  account.  (Note: also prepared support for JFS2 quota classes; this is
  yet unfinished and untested. See sub-directory contrib/aix_jfs2_class)
- Added support for XFS project quotas on request of
  Sten Spans (sten@blinkenlights.nl); received no feedback it if works.
  Changed the "isgrp" param to Quota::query() and setqlim() from boolean
  into an enum: 0:=user, 1:=group, 2:=project quotas.
- Added a copyright notice to the main module and the manual page on
  request of the Debian maintainer.

Changes in 1.5.1 (July 2005)
- Added new API function rpcauth: use non-default authentication for RPC
  Patch submitted by Karol Lassak
- Makefile fix for Linux: compile linuxapi.c with default CFLAGS
  because for 64-bit systems -fPIC is required.  Patches submitted
  by David Lee and James Olin Oden; used a different solution though.

Changes in 1.5.0 (Nov 2004)
- Added new API function rpcpeer: set port and protocol for RPC queries.
  On request by Sten Spans (sten@blinkenlights.nl)

Changes in 1.4.11 (Aug 2004)
- Updated VxFS detection in Makefile.PL for recent Solaris/VxFS versions
  Thanks to Joshua Frigerio for the info and to Keith Clay for verification.
- tiny syntax cleanup in linuxapi.c

Changes in 1.4.10 (Jan 2004)
- Applied patch by Antti Tapaninen (aet at cc.hut.fi) to add support
  for MacOS X (based on BSD config) and NFS3 support for AIX.
- Applied patch by Chris Adams (cmadams at hiwaay.net) to avoid warning
  in Quota::getqcarg() on systems where getmntent(3) may return invalid
  mount points, e.g. Tru64 UNIX

Changes in 1.4.9 (Aug 2003)
- Applied patch by Wolfgang.Friebel@desy.de: Changed AFS quota support,
  now based on OpenAFS instead of arla and Kerberos4.

Changes in 1.4.8 (May 2003)
- fixed bug in linuxapi.c, thanks to Shane DeRidder

Changes in 1.4.7 (May 2003)
- adapted for the latest fashion of Linux quota APIs (version 6.5.1)
  Thanks to Jay R. Wren, Shane DeRidder, Roman "romke" and Geoffrey Pan
  for your infos and feedback.
- added quota threshold monitoring script to the contrib directory.
  Thanks to Ziaur Rahman.

Changes in 1.4.6 (August 2002)
- changed the test script to exit if STDIN is not a terminal device.
  This is an interactive script which cannot be run by CPANPLUS.

Changes in 1.4.5 (July 2002)
- replaced use of macro BCOPY with libc function memcpy() because of build
  problems on Solaris 2.8 (bcopy is depreciated now anyways)
  Thanks to Jost Krieger and Paul Sand for reporting this problem.

Changes in 1.4.4 (June 2002)
- bugfix in Quota::getqcarg(): the function failed if the device id returned
  by stat(2) for the queried path was zero (e.g. for the root fs on NetBSD)
  Thanks to Jake Fan (jake@chaogic.com) for reporting this bug.

Changes in 1.4.3 (May 2002)
- updated for new version of Solaris VxFS (Veritas Filesystem)
  by David Lee (T.D.Lee@durham.ac.uk)

Changes in 1.4.2 (Jan. - Mar. 2002)
- fixed test.pl to allow uids with more than 5 digits.
  Thanks to Neil Prockter (perl@iamafreeman.com) for the fix.
- updated Linux quota version detection with quota-tools 3.04;
  removed compile-time version detection because it caused problems;
  added switch LINUX_API_VERSION to hints/linux.h to allow to hard-wire
  the API in case the automatic version detection fails.

Changes in 1.4.1 (Sep. 2001 - Jan. 2002)
- added support for an older (intermediate?) version of the new Linux Quota
  API which contains a mixture of v1 and v2 command ids (uses the old GETSTAT

Quota/Quota.pm
# ------------------------------------------------------------------------ #
# Quota.pm - Copyright (C) 1995-2013 Tom Zoerner
# ------------------------------------------------------------------------ #
# This program is free software: you can redistribute it and/or modify
# it either under the terms of the Perl Artistic License or the GNU
# General Public License as published by the Free Software Foundation.
# (Either version 2 of the GPL, or any later version.)
# For a copy of these licenses see <http://www.opensource.org/licenses/>.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# Perl Artistic License or GNU General Public License for more details.
# ------------------------------------------------------------------------ #

package Quota;

require Exporter;
use AutoLoader;
require DynaLoader;

@ISA = qw(Exporter DynaLoader);
@EXPORT = ();

$VERSION = '1.7.1';

bootstrap Quota;

use Carp;
use POSIX qw(:errno_h);
use strict;

##
##  Get block device for locally mounted file system
##  !! Do not use this to get the argument for the quota-functions in this
##  !! module, since not all operating systems use the device path for the
##  !! quotactl system call and e.g. Solaris doesn't even use a system call
##  !! Always use getqcarg() instead.
##

sub getdev {
  ($#_ > 0) && croak("Usage: Quota::getdev(path)");
  my($target) = (($#_ == -1) ? "." : $_[0]);
  my($dev) = (stat($target))[0];
  my($ret) = undef;
  my($fsname,$path);
 
  if($dev && ($target ne "") && !Quota::setmntent()) {
    while(($fsname,$path) = Quota::getmntent()) {
      ($ret=$fsname, last) if ($dev == (stat($path))[0]);
    }
    $! = 0;
  }
  Quota::endmntent();
  $ret;
}

##
##  Get "device" argument for this module's Quota-functions
##

sub getqcarg {
  ($#_ > 0) && croak("Usage: Quota::getqcarg(path)");
  my($target) = (($#_ == -1) ? "." : $_[0]);
  my($dev) = (stat($target))[0];
  my($ret) = undef;
  my($argtyp,$fsupp) = (Quota::getqcargtype() =~ /([^,]*)(,.*)?/);
  my($fsname,$path,$fstyp,$fsopt);

  if(defined($dev) && ($target ne "") && !Quota::setmntent()) {
    while(($fsname,$path,$fstyp,$fsopt) = Quota::getmntent()) {
      next if $fstyp =~ /^(lofs|ignore|auto.*|proc|rootfs)$/;
      my($pdev) = (stat($path))[0];
      if (defined($pdev) && ($dev == $pdev)) {
        if ($fsname =~ m|^[^/]+:/|) {
          $ret = $fsname;  #NFS host:/path
        } elsif (($fstyp =~ /^nfs/i) && ($fsname =~ m#^(/.*)\@([^/]+)$#)) {
          $ret = "$2:$1";  #NFS /path@host
        } elsif ($argtyp eq "dev") {
          if ($fsopt =~ m#(^|,)loop=(/dev/[^,]+)#) {
            $ret = $2;  # Linux mount -o loop
          } else {
            $ret = $fsname;
          }
        } elsif ($argtyp eq "qfile") {
          $ret = "$path/quotas";
        } elsif ($argtyp eq "any") {
          $ret = $target;
        } else { #($argtyp eq "mntpt")
          $ret = $path;
        }

        # XFS, VxFS and AFS quotas require separate access methods
        # (optional for VxFS: later versions use 'normal' quota interface)
        if   (($fstyp eq "xfs") && ($fsupp =~ /,XFS/)) { $ret = "(XFS)$ret" }
        elsif(($fstyp eq "vxfs") &&
              defined($fsupp) && ($fsupp =~ /,VXFS/)) { $ret = "(VXFS)$ret" }
        elsif((($fstyp eq "afs") || ($fsname eq "AFS")) &&
              ($fsupp =~ /,AFS/)) { $ret = "(AFS)$target"; }
        if   (($fstyp eq "jfs2") && ($fsupp =~ /,JFS2/)) { $ret = "(JFS2)$ret" }
        last;
      }
    }
    $! = 0;
  }
  Quota::endmntent();
  $ret;
}

##
##  Translate error codes of quotactl syscall and ioctl
##

sub strerr {
  ($#_ != -1) && croak("Usage: Quota::strerr()");
  my($str);

  eval {
    if(($! == &EINVAL) || ($! == &ENOTTY) || ($! == &ENOENT) || ($! == ENOSYS))
                         { $str = "No quotas on this system" }
    elsif($! == &ENODEV) { $str = "Not a standard file system" }
    elsif($! == &EPERM)  { $str = "Not privileged" }
    elsif($! == &EACCES) { $str = "Access denied" }
    elsif($! == &ESRCH)  { $str = "No quota for this user" }
    elsif($! == &EUSERS) { $str = "Quota table overflow" }
    else { die "unknown quota error\n" }
  };
  if($@) {
    my($err) = $! + 0;
    $str = "error #$err";
  };
  $str;
}

package Quota; # return to package Quota so AutoSplit is happy
1;
__END__

=head1 NAME

Quota - Perl interface to file system quotas

=head1 SYNOPSIS

    use Quota;

    ($block_curr, $block_soft, $block_hard, $block_timelimit,
     $inode_curr, $inode_soft, $inode_hard, $inode_timelimit) =
    Quota::query($dev [,$uid [,kind]]);

    ($block_curr, $block_soft, $block_hard, $block_timelimit,
     $inode_curr, $inode_soft, $inode_hard, $inode_timelimit) =
    Quota::rpcquery($host, $path [,$uid [,kind]]);

    Quota::rpcpeer([$port [,$use_tcp [,timeout]]]);
    
    Quota::rpcauth([$uid [,$gid [,$hostname]]]);

    Quota::setqlim($dev, $uid, $block_soft, $block_hard,
                   $inode_soft, $inode_hard [,$tlo [,kind]]);

    Quota::sync([$dev]);

    $arg = Quota::getqcarg([$path]);

    Quota::setmntent();
    ($dev, $path, $type, $opts) = Quota::getmntent();
    Quota::endmntent();

=head1 DESCRIPTION

The B<Quota> module provides access to file system quotas.
The quotactl system call or ioctl is used to query or set quotas
on the local host, or queries are submitted via RPC to a remote host.
Mount tables can be parsed with B<getmntent> and paths can be
translated to device files (or whatever the actual B<quotactl>
implementations needs as argument) of the according file system.

=head2 Functions

=over 4

=item I<($bc,$bs,$bh,$bt, $ic,$is,$ih,$it) = Quota::query($dev, $uid, $kind)>

Get current usage and quota limits for a given file system and user.
The user is specified by its numeric uid; defaults to the process'
real uid.

The type of I<$dev> varies from system to system. It's the argument
which is used by the B<quotactl> implementation to address a specific
file system. It may be the path of a device file (e.g. B</dev/sd0a>)
or the path of the mount point or the quotas file at the top of
the file system (e.g. B</home.stand/quotas>). However you do not
have to worry about that; use B<Quota::getqcarg> to automatically
translate any path inside a file system to the required I<$dev> argument.

I<$dev> may also be in the form of B<hostname:path>, which has the
module transparently query the given host via a remote procedure call
(RPC). In case you have B<NFS> (or similar network mounts), this type
of argument may also be produced by B<Quota::getqcarg>. Note: RPC
queries require I<rquotad(1m)> to be running on the target system. If
the daemon or host are down, the timeout is 12 seconds.

In I<$bc> and I<$ic> the current usage in blocks and inodes is returned.
I<$bs> and I<$is> are the soft limits, I<$bh> and I<$ih> hard limits. If the
soft limit is exceeded, writes by this user will fail for blocks or
inodes after I<$bt> or I<$it> is reached. These times are expressed
as usual, i.e. in elapsed seconds since 00:00 1/Jan/1970 GMT.

Note: When the quota limits are not exceeded, the timestamps
are meaningless and should be ignored. When hard and soft
limits are zero, there is no limit for that user. On most
systems Quota::query will return undef in that case and
errno will be set to ESRCH.

When I<$kind> is given and set to 1, I<$uid> is taken as gid and
group quotas are queried. This is B<not> supported across RPC and
even locally only on a few architectures (e.g. Linux and other BSD
based Unix variants, OSF/1 and  AIX - check the quotactl(2) man page
on your systems). When I<$kind> is set to 2, project quotas are
queried; this is currently only supported for XFS.  When unsupported,
this flag is ignored.

=item I<Quota::setqlim($dev, $uid, $bs,$bh, $is,$ih, $tlo, $kind)>

Sets quota limits for the given user. Meanings of I<$dev>, I<$uid>,
I<$bs>, I<$bh>, I<$is> and I<$ih> are the same as in B<Quota::query>.

For file systems exceeding 2 TB: To allow passing block or inode
values larger or equal to 2^32 on 32-bit Perl versions, pass them
either as strings or floating point.

I<$tlo> decides how the time limits are initialized:
I<0>: The time limits are set to B<NOT STARTED>, i.e. the time limits
are not initialized until the first write attempt by this user.
This is the default.
I<1>: The time limits are set to B<7.0 days>.
More alternatives (i.e. setting a specific time) aren't available in
most implementations.

When I<$kind> is given and set to 1, I<$uid> is taken as gid and
group quota limits are set. This is supported only on a few
architectures (see above). When I<$kind> is set to 2, project
quotas are modified; this is currently only supported for XFS.
When unsupported, this flag is ignored.

Note: if you want to set the quota of a particular user to zero, i.e.
no write permission, you must not set all limits to zero, since that
is equivalent to unlimited access. Instead set only the hard limit
to 0 and the soft limit for example to 1.

Note that you cannot set quotas via RPC.

=item I<Quota::sync($dev)>

Have the kernel update the quota file on disk or all quota files
if no argument given (the latter doesn't work on all systems,
in particular on B<HP-UX 10.10>).

The main purpose of this function is to check if quota is enabled
in the kernel and for a particular file system. Read the B<quotaon(1m)>
man page on how to enable quotas on a file system.

Note: on some systems this function always returns a success indication,
even on partitions which do not have quotas enabled (e.g. Linux 2.4).
This is not a bug in this module; it's a limitation in certain kernels.

=item I<($bc,$bs,$bh,$bt, $ic,$is,$ih,$it) =>

I<Quota::rpcquery($host,$path,$uid,$kind)>

This is equivalent to B<Quota::query("$host:$path",$uid,$kind)>, i.e.
query quota for a given user on a given remote host via RPC.
I<$path> is the path of any file or directory inside the
file system on the remote host.  Querying group quotas ($kind = 1)
is only recently supported on some platforms (e.g. on linux via
"extended" quota RPC, i.e. quota RPC version 2) so it may fail due
to lack of support either on client or server side, or both.

=item I<Quota::rpcpeer($port,$use_tcp,timeout)>

Configure parameters for subsequent RPC queries; all parameters are
optional.  By default the portmapper on the remote host is used
(i.e. default port is 0, protocol is UDP)  The default timeout is
4 seconds.

=item I<Quota::rpcauth($uid,$gid,$hostname)>

Configure authorization parameters for subsequent RPC queries; 
all parameters are optional. By default uid and gid are taken from 
owner of the process and hostname is the host name of current machine.

=item I<$arg = Quota::getqcarg($path)>

Get the required I<$dev> argument for B<Quota::query> and B<Quota::setqlim>
for the file system you want to operate on. I<$path> is any path of an
existing file or directory inside that file system. The path argument is
optional and defaults to the current working directory.

The type of I<$dev> varies between operating systems, i.e. different
implementations of the quotactl functionality. Hence it's important for
compatibility to always use this module function and not really pass
a device file to B<Quota::query> (as returned by B<Quota::getdev>).
See also above at I<Quota::query>

=item I<$dev = Quota::getdev($path)>

Returns the device entry in the mount table for a particular file system,
specified by any path of an existing file or directory inside it. I<$path>
defaults to the working directory. This device entry need not really be
a device. For example on network mounts (B<NFS>) it's I<"host:mountpath">,
with I<amd(1m)> it may be something completely different.

I<NEVER> use this to produce a I<$dev> argument for other functions of
this module, since it's not compatible. On some systems I<quotactl>
does not work on devices but on the I<quotas> file or some other kind of
argument. Always use B<Quota::getqcarg>.

=item I<Quota::setmntent()>

Opens or resets the mount table. This is required before the first
invocation of B<Quota::getmntent>.

Note: on some systems there is no equivalent function in the C library.
But you still have to call this module procedure for initialization of
module-internal variables.

=item I<($dev, $path, $type, $opts) = Quota::getmntent()>

Returns the next entry in the system mount table. This table contains
information about all currently mounted (local or remote) file systems.
The format and location of this table (e.g. B</etc/mtab>) vary from
system to system. This function is provided as a compatible way to
parse it. (On some systems, like B<OSF/1>, this table isn't
accessible as a file at all, i.e. only via B<Quota::getmntent>).

=item I<Quota::endmntent()>

Close the mount table. Should be called after the last use of
B<Quota::getmntent> to free possibly allocated file handles and memory.
Always returns undef.

=item I<Quota::strerr()>

Translates B<$!> to a quota-specific error text. You should always
use this function to output error messages, since the normal messages
don't always make sense for quota errors
(e.g. I<ESRCH>: B<No such process>, here: B<No quota for this user>)

Note that this function only returns a defined result if you called a
Quota command directly before which returned an error indication.

=back

=head1 RETURN VALUES

Functions that are supposed return lists or scalars, return I<undef> upon
errors. As usual B<$!> contains the error code (see B<Quota::strerr>).

B<Quota::endmntent> always returns I<undef>.
All other functions return 0 upon success, non-zero integer otherwise.

=head1 EXAMPLES

An example for each function can be found in the test script
I<test.pl>. See also the contrib directory, which contains
some longer scripts, kindly donated by users of the module.

=head1 BUGS

With remote quotas we have to rely on the remote system to
state correctly which block size the quota values are
referring to. Old versions of the Linux rpc.rquotad
reported a block size of 4 kilobytes, which was wildly
incorrect. For more info on this and other Linux bugs please
see INSTALL.

=head1 AUTHORS

This module was created 1995 by Tom Zoerner
(email: tomzo AT users.sourceforge.net)
and since then continually improved and ported to
many operating- and file-systems. Numerous people
have contributed to this process; for a complete
list of names please see the CHANGES document.

The quota module was in the public domain 1995-2001. Since 2001 it is
licensed under both the Perl Artistic License and version 2 or later of the
GNU General Public License as published by the Free Software Foundation.
For a copy of these licenses see <http://www.opensource.org/licenses/>.
The respective authors of the source code are it's owner in regard to
copyright.

=head1 SEE ALSO

perl(1), edquota(1m),
quotactl(2) or quotactl(7I),
mount(1m), mtab(4) or mnttab(4), quotaon(1m),
setmntent(3), getmntent(3) or getmntinfo(3), endmntent(3),
rpc(3), rquotad(1m).

=cut
